### ISSUE TITLE
Missing node_name in chef-client config causing wrong node_name used

### ISSUE THREAD

**Issue by heaven** [chef-cookbooks/chef-client#536]
Chef client doesn't add `node_name` to `chef.rb`, I'm wondering if this such a big optimization? 

This simply doesn't work:
```ruby
'node_name' => Chef::Config[:node_name] == node['fqdn'] ? false : Chef::Config[:node_name],
```

Because we use custom hostnames, our local host name is `monitor` while the node is named using FQDN value `monitor.nyc3.***.com`

Our `/etc/hosts`:
```
127.0.0.1	localhost
127.0.1.1	monitor
10.132.***.***	monitor.nyc3.***.com graphite.nyc3.***.com monitor.devops graphite.devops
ff02::3	ip6-allhosts
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
::1	ip6-localhost ip6-loopback
fe00::	ip6-localnet
ff00::	ip6-mcastprefix
```

As a result, we receive:
```
root@monitor:~# chef-client
Starting Chef Client, version 12.21.26

================================================================================
Chef encountered an error attempting to load the node data for "monitor"
================================================================================

Authentication Error:
---------------------
Failed to authenticate to the chef server (http 401).

Server Response:
----------------
Failed to authenticate as 'monitor'. Ensure that your node_name and client key are correct.

Relevant Config Settings:
-------------------------
chef_server_url   "https://chef-1.***.com/organizations/wego"
node_name         "monitor"
client_key        "/etc/chef/client.pem"

If these settings are correct, your client_key may be invalid, or
you may have a chef user with the same client name as this node.

System Info:
------------
chef_version=12.21.26
ruby=ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux]
program_name=chef-client worker: ppid=17044;start=14:33:11;
executable=/opt/chef/bin/chef-client
```

**Comment by msgarbossa** [comment]
I agree node_name should always be written to client.rb.  I've noticed knife bootstrap has more prominently enforced setting node_name in the last few years.  The node_name should <b>never</b> change and shouldn't be left to an arbitrary interpretation of the FQDN or else we end up with authentication issues like these.  Having node_name in client.rb also allows knife from the chef-client package to work by passing -c /etc/chef/client.rb to any commands.  This code in the chef-client cookbook appears to be about 5 years old.  Maybe this idea of excluding node_name should be revisited.  I can't see the downside to having it there.  Minimally a simpler way of forcing Chef::Config[:node_name] to be written would be an improvement.  This issue is similar to #493, which suggests setting a normal attribute.  I took a slightly different approach and put the following line in the wrapper cookbook, which overrides the default attribute with the same test condition that sets it to false.

```
node.default['chef_client']['config']['node_name'] = Chef::Config[:node_name]
  if Chef::Config[:node_name] == node['fqdn']
include_recipe 'chef-client::config'
```

**Comment by tas50** [comment]
Fixed in 10.0

**Comment by Tyrael** [comment]
any chance this coul be updated in the docs also? (https://github.com/chef/chef-web-docs/blob/master/chef_master/source/config_rb.rst) currently it suggests to omit the node_name from the client conf if it equals to the fqdn (which is wrong because everything but chef-client and chef-shell will complain about the missing client name/node_name)
